### PR TITLE
[Qwen3] Switch to verified RoPE implementation + Add weight tying support

### PR DESCRIPTION
**Pull request by wwwjn**
## Context
1. Current Qwen3 model RoPE used some trick to make numerical parity with HF. This trick is from un-official source and hard to reasoning mathematically. Switch to [torchtune based implementation](https://github.com/pytorch/torchtune/blob/main/torchtune/models/qwen2/_positional_embeddings.py#L14), which was directly contributed from Qwen team.  Thanks @ebsmothers for point us to this implementation!
    - For RoPE embedding, I change it to the same way as complex representation based RoPE in llama3: We initialize and precompute the RoPE embedding cos/sin value only once, and pass it into Attention module during forward. In this way, TP can be applied seamlessly.
    - In contrast, torchtune passed the RoPE class into initialize function for each layers' attention module.
2. Add weight tying support for Qwen3, verified with FSDP + TP

## Numerical verification for RoPE
Run end-to-end forward pass of Qwen3 model, the output and 
Screenshot 2025-08-18 at 2 48 48 PM

## Weight tying Verification:

1. With vs. without weight tying on torchtitan model: (FSDP=4, loss are exactly the same)
Screenshot 2025-08-18 at 6 19 13 PM

2. torchtitan with weight tying vs. HF  
Screenshot 2025-08-18 at 9 37 50 PM

3. Weight tying memory address / id check: (in train.py) - passed
```
assert id(model.tok_embeddings.weight) == id(model.output.weight), "id check 2"
assertEqual(model.tok_embeddings.weight, model.output.weight)

# model.forward()

assert id(model.tok_embeddings.weight.grad) == id(model.output.weight.grad), "id check 2"
assertEqual(model.tok_embeddings.weight.grad, model.output.weight.grad)
```

### CODE REVIEW

File: torchtitan/models/llama3/infra/parallelize.py [hunk 1]
```diff
@@ -362,6 +362,7 @@ def apply_fsdp(
     pp_enabled: bool,
     cpu_offload: bool = False,
     reshard_after_forward_policy: str = "default",
+    enable_weight_tying: bool = False,
```
Comment (tianyu-l) [inline torchtitan/models/llama3/infra/parallelize.py]: Maybe I missed, but who calls this today?
Also, I think we may need to have a separate `apply_fsdp`, as llama3 doesn't have this option. We can probably remove PP-related stuff there.

File: torchtitan/experiments/qwen3/model/model.py [hunk 2]
```diff
@@ -370,13 +348,22 @@ def __init__(self, model_args: Qwen3ModelArgs):
         self.head_dim = model_args.head_dim
 
         self.tok_embeddings = nn.Embedding(model_args.vocab_size, model_args.dim)
-        self.register_buffer("freqs_cis", self._precompute_freqs_cis(), persistent=True)
+
+        # Set persistent=False for now because Qwen3 Model hasn't supported Pipeline Parallel yet.
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/model/model.py]: PP doesn't need this to be True any more. We can test if compile is happy with it.

Comment (wwwjn) [reply torchtitan/experiments/qwen3/model/model.py]: Great! I've tested compile works fine with this

File: torchtitan/experiments/qwen3/model/state_dict_adapter.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                abstract_key = re.sub(r"(\d+)", "{}", key, count=1)
+                layer_num = re.search(r"\d+", key).group(0)
+                new_key = to_hf_map[abstract_key]
+                # We need to permute the weights in wq and wk layer in order to account for the difference between
+                # the native Llama and huggingface RoPE implementation.
+                # if abstract_key == "layers.{}.attention.wq.weight":
+                #     value = self._permute(value, n_heads)
+                # if abstract_key == "layers.{}.attention.wk.weight":
+                #     key_value_dim = head_dim * n_kv_heads
+                #     value = self._permute(value, n_kv_heads, key_value_dim, dim)
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/model/state_dict_adapter.py]: need to permute or not?

File: torchtitan/experiments/qwen3/model/model.py [hunk 4]
```diff
@@ -370,13 +348,18 @@ def __init__(self, model_args: Qwen3ModelArgs):
         self.head_dim = model_args.head_dim
 
         self.tok_embeddings = nn.Embedding(model_args.vocab_size, model_args.dim)
-        self.register_buffer("freqs_cis", self._precompute_freqs_cis(), persistent=True)
+
+        self.register_buffer(
+            "rope_cache", self._precompute_rope_cache(), persistent=False
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/model/model.py:353]: If it doesn't need to be persistent for compile to work. Let's also remove them universally, in Llama 3, Llama 4, DSV3, and this field in general https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/checkpoint.py#L61.

Let's test and do this in a separate PR.

File: torchtitan/experiments/qwen3/model/state_dict_adapter.py (hunk 3 above)
Comment (wwwjn) [reply torchtitan/experiments/qwen3/model/state_dict_adapter.py]: No need to permute, with updated RoPE 
- Rotated weights  + Eleuther RoPE + don't permute weights.

File: torchtitan/experiments/qwen3/model/model.py (hunk 4 above)
Comment (wwwjn) [reply torchtitan/experiments/qwen3/model/model.py:353]: > Let's test and do this in a separate PR.

Sure, adding it

**Review by tianyu-l — APPROVED** [review]
LGTM.
I saw you removed the state dict adapter. Do we plan to add it back?

**Comment by wwwjn** [comment]
> LGTM. I saw you removed the state dict adapter. Do we plan to add it back?

Yes, I will let @HosseinKaviani-H  add it back cause he implemented the statedictAdapter, and I borrowed it for testing purpose.
